### PR TITLE
fix: skipping label if it contains special symbol

### LINKED ISSUE #13963: [loki] Json|Logfmt Parser returns empty label value if it contains symbol `�`
**Describe the bug**

When the Logfmt or JSON parser meets the symbol `�` inside the label value, it skips the whole label or replaces the value with an empty string.  

Loki drops the value([here](https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/parser.go#L205) and [here](https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/parser.go#L343)) because PromQL engine fails the query if it meets such symbols. 

**To Reproduce**
Open LogQL analyzer page and run the query:

1. [Logfmt example](https://grafana.com/docs/loki/latest/query/analyzer/?query=%7C+logfmt+&line%5B%5D=level%3Dinfo+ts%3D2022-03-23T11%3A55%3A29.846163306Z+caller%3Dmain.go%3A112+msg%3D%22Starting+%EF%BF%BD+Enterprise+Logs%22)
image

2. [JSON example](https://grafana.com/docs/loki/latest/query/analyzer/?query=%7C+json+&line%5B%5D=%7B%22timestamp%22%3A%222022-04-26T08%3A53%3A59.61Z%22%2C%22level%22%3A%22INFO%22%2C%22message%22%3A%22Starting+%EF%BF%BD+Enterprise+Logs%22%7D)
image

**Expected behavior**
Label value must be replaced with ` `(symbol: space) instead of dropping the label or label value. 

**Environment:**
 - Infrastructure: any
 - Deployment tool: any

### PR DESCRIPTION
**Pull request by ravishankar15**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #13963 

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [x] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [x] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/logql/log/parser.go [hunk 1]
```diff
@@ -200,12 +201,16 @@ func unescapeJSONString(b []byte) string {
...
 	}
 	res := string(bU)
-	// rune error is rejected by Prometheus
-	for _, r := range res {
+
+	// rune error is rejected by Prometheus hence replacing them with space
+	removeInvalidUtf := func(r rune) rune {
 		if r == utf8.RuneError {
-			return ""
+			return -1
```
Comment (vlad-diachenko) [inline pkg/logql/log/parser.go:208]: it's necessary to replace it with ` ` (whitespace) instead of skipping it

File: pkg/logql/log/parser.go [hunk 2]
```diff
@@ -200,12 +201,16 @@ func unescapeJSONString(b []byte) string {
...
 		}
+		return r
 	}
+	res = strings.Map(removeInvalidUtf, res)
```
Comment (vlad-diachenko) [inline pkg/logql/log/parser.go:212]: see:  https://github.com/grafana/loki/pull/14068#issuecomment-2337504002
```suggestion
	if strings.ContainsRune(source, utf8.RuneError) {
		res = strings.Map(removeInvalidUtf, res)
	}
```

File: pkg/logql/log/parser.go [hunk 3]
```diff
@@ -200,12 +201,16 @@ func unescapeJSONString(b []byte) string {
 		return ""
 	}
 	res := string(bU)
-	// rune error is rejected by Prometheus
-	for _, r := range res {
+
+	// rune error is rejected by Prometheus hence replacing them with space
+	removeInvalidUtf := func(r rune) rune {
```
Comment (vlad-diachenko) [inline pkg/logql/log/parser.go:206]: nit: please move this function outside of `unescapeJSONString` function to re-use it in logfmt parser ...

File: pkg/logql/log/parser.go [hunk 4]
```diff
@@ -339,10 +344,15 @@ func (l *LogfmtParser) Process(_ int64, line []byte, lbs *LabelsBuilder) ([]byte
 		}
 
 		val := l.dec.Value()
-		// the rune error replacement is rejected by Prometheus, so we skip it.
-		if bytes.ContainsRune(val, utf8.RuneError) {
-			val = nil
+
+		// the rune error replacement is rejected by Prometheus hence skiping them.
+		removeInvalidUtf := func(r rune) rune {
```
Comment (vlad-diachenko) [inline pkg/logql/log/parser.go:349]: nit: can be re-used from json parse

File: pkg/logql/log/parser.go [hunk 5]
```diff
@@ -339,10 +344,15 @@ func (l *LogfmtParser) Process(_ int64, line []byte, lbs *LabelsBuilder) ([]byte
...
+			if r == utf8.RuneError {
+				return -1
+			}
+			return r
 		}
+		val = bytes.Map(removeInvalidUtf, val)
```
Comment (vlad-diachenko) [inline pkg/logql/log/parser.go:355]: see: https://github.com/grafana/loki/pull/14068#issuecomment-2337504002
```suggestion
		if bytes.ContainsRune(source, utf8.RuneError) {
			val = bytes.Map(removeInvalidUtf, val)
		}
```

**Review by vlad-diachenko — CHANGES_REQUESTED** [review]
looks great, however, we need to address some small things to get it merged

**Comment by ravishankar15** [comment]
@vlad-diachenko Thanks for the insights :) I have made the suggestion

**Review by vlad-diachenko — APPROVED** [review]
looks awesome 💎 🥇 huge thanks for the contribution @ravishankar15

**Pull request merged**